### PR TITLE
Update TestActions.plp for new relative flag behavior and trench grind lengthened

### PR DESCRIPTION
**Pull request by AstroStucky**
## Jira Ticket
[OW-1154](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1154)

## Summary of Changes
* [OW-1121](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1121) Changed the behavior of the relative flag. This PR updates TestActions.plp to reflect that change in parameter behavior.
* The commanded TaskScoopLinear would push the arm farther than the length of the trench carved out by the grinder, causing it to jam at the far end.  The length of the grind has been increased to fix this problem. 

## Test
Launch any world and call the following tests in this order (so long as TestActions.plp comes last these will work in the same simulation instance). 
1. TestPanTiltCartesian.plp 
2. TestArmMoveCartesianGuarded.plp
3. TestActions.plp

## Merge Conflict Risk
This can be expected to cause a merge conflict when Release 11.1 is merged back into noetic-devel, but given the changes are small the conflict should be simple to resolve.

### CODE REVIEW

**Comment by kmdalal** [comment]
The first `ArmMoveCartesian` consistently produces this warning:
```
[INFO]: ArmMoveCartesian action started
[WARN]: Orientation is not normalized and will be interpreted as (0.15796380630708962, 0.2657169736006032, 0.7336112916224276, 0.6051912083313722) instead of the provided value
```
I don't understand it since the orientation sent is (0,0,0).

**Comment by kmdalal** [comment]
The `DockIngestSample` did not complete:
```
[INFO]: DockIngestSample action started

libcurl: (28) Operation timed out after 300121 milliseconds with 0 out of 0 bytes received
```
I've not seen this before and suspect it was a rare glitch of some kind.  I don't know why `libcurl` is involved here.

**Comment by kmdalal** [comment]
I'd like to extend this PR to fix all plans broken by the given update.  In ReferenceMission1, the circular digs fail:
```
[ERROR]: TaskScoopCircular: Aborted - The follow_joint_trajectory action server for arm_controller resulted in error code -5 with the message: j_dist_pitch goal error 0.698875
```
and later:
```
[ERROR]: TaskScoopCircular: Aborted - The follow_joint_trajectory action server for arm_controller resulted in error code -5 with the message: j_dist_pitch goal error 0.698778
```

**Comment by kmdalal** [comment]
Also in RM1, a discard fails:
```
[INFO]: TaskDiscardSample action started
[ WARN]: Solution is approximate. This usually indicates a failure.
[ WARN]: Fail: ABORTED: No motion plan found. No execution attempted.
[ERROR]: TaskDiscardSample: Aborted - MoveIt planning failed with error code: val: -1
```

**Comment by kmdalal** [comment]
The plan `TestArmMoveCartesianGuarded` has a number of failures and (@Lumsterling) also exposes the excessive power warning:
```
[INFO]: ArmMoveCartesian action started
[ WARN]: Power system node computed excessive power input for GSAP, 62.1105W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 62.2145W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 82.6193W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 82.5862W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 82.7938W. Capping GSAP input at 20W.
[ERROR]: ArmMoveCartesian: Aborted - Failed to reach intended pose
[INFO]: ArmMoveCartesian action complete
[INFO]: ArmMoveCartesianGuarded action started
[ WARN]: Power system node computed excessive power input for GSAP, 83.0699W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 83.2844W. Capping GSAP input at 20W.
[INFO]: ArmMoveCartesianGuarded: Succeeded - ArmMoveCartesianGuarded trajectory completed without breaching force or torque thresholds
[INFO]: ArmMoveCartesianGuarded action complete
[INFO]: ArmMoveCartesian action started
[ WARN]: Power system node computed excessive power input for GSAP, 83.6472W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 83.8982W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 83.9008W. Capping GSAP input at 20W.
[INFO]: ArmMoveCartesian: Succeeded - ArmMoveCartesian trajectory succeeded
[INFO]: ArmMoveCartesian action complete
[INFO]: ArmMoveCartesianGuarded action started
[ WARN]: Power system node computed excessive power input for GSAP, 83.9623W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.0228W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.0771W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.1153W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.1153W. Capping GSAP input at 20W.
[INFO]: ArmMoveCartesianGuarded: Succeeded - ArmMoveCartesianGuarded trajectory completed without breaching force or torque thresholds
[INFO]: ArmMoveCartesianGuarded action complete
[INFO]: ArmUnstow action started
[ERROR]: Got a transition callback on a goal handle that we're not tracking
[ WARN]: Power system node computed excessive power input for GSAP, 84.3251W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.5406W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.7538W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.754W. Capping GSAP input at 20W.
[INFO]: ArmUnstow: Succeeded - ArmUnstow trajectory succeeded
[INFO]: ArmStow action started
[INFO]: ArmUnstow action complete
[ WARN]: Power system node computed excessive power input for GSAP, 84.8758W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.8811W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.8709W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.8628W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.7936W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 84.6292W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 24.8433W. Capping GSAP input at 20W.
[ WARN]: Power system node computed excessive power input for GSAP, 24.6201W. Capping GSAP input at 20W.
[INFO]: ArmStow: Succeeded - ArmStow trajectory succeeded
[INFO]: ArmStow action complete
```
I witnessed exactly when the first excessive power warning occurred: it was when the arm was stuck in the ground, and this was due to a wanky orientation and position of the arm.  Apparently the excessive power draw continued even after the arm was unstuck, and stowing.

**Comment by AstroStucky** [comment]
> The first `ArmMoveCartesian` consistently produces this warning:
> 
> ```
> [INFO]: ArmMoveCartesian action started
> [WARN]: Orientation is not normalized and will be interpreted as (0.15796380630708962, 0.2657169736006032, 0.7336112916224276, 0.6051912083313722) instead of the provided value
> ```
> 
> I don't understand it since the orientation sent is (0,0,0).

I think I found the bug. 

OwInterface.cpp:1210
```C++
  q.setEuler (z,y,x);  // Yaw, pitch, roll, respectively.
```
should be 
```C++
  q.setEuler (orient_z,orient_y,orient_x);  // Yaw, pitch, roll, respectively.
```

**Comment by AstroStucky** [comment]
> The `DockIngestSample` did not complete:
> 
> ```
> [INFO]: DockIngestSample action started
> 
> libcurl: (28) Operation timed out after 300121 milliseconds with 0 out of 0 bytes received
> ```
> 
> I've not seen this before and suspect it was a rare glitch of some kind. I don't know why `libcurl` is involved here.

Unfortunately there is some randomness to the DockIngestSample action that is mostly due to randomness in how regolith behaves in a given simulation. For example my most recent run resulted in
```
[INFO:/lander_action_servers] DockIngestSample action started
[ERROR:/lander_action_servers] DockIngestSample: Aborted - Regolith removal service failed
[INFO:/lander_action_servers] DockIngestSample action complete
```
I suspect it has something to do with a regolith that didn't make it into the sample dock. Regardless, you should have still seen `[INFO:/lander_action_servers] DockIngestSample action complete` eventually. Did you never see this?

**Comment by kmdalal** [comment]
> I suspect it has something to do with a regolith that didn't make it into the sample dock. Regardless, you should have still seen `[INFO:/lander_action_servers] DockIngestSample action complete` eventually. Did you never see this?

Yes, and it's in the output I showed.  However, for some reason the PLEXIL plan got hung on that action.

**Comment by kmdalal** [comment]
> I think I found the bug.
> 
> OwInterface.cpp:1210

Wow, good catch!  And that probably got into Release 11.  After this is tested I'll create a hotfix.

**Comment by kmdalal** [comment]
> > I think I found the bug.
> > OwInterface.cpp:1210

The same silly bug occurs for the Guarded version of this action on line 1243.  I suggest fixing these on this branch -- either of us can commit it.

**Comment by AstroStucky** [comment]
> > I suspect it has something to do with a regolith that didn't make it into the sample dock. Regardless, you should have still seen `[INFO:/lander_action_servers] DockIngestSample action complete` eventually. Did you never see this?
> 
> Yes, and it's in the output I showed. However, for some reason the PLEXIL plan got hung on that action.

That is definitely not normal behavior for DockIngestSample. I'll make a ticket for this and the error I witnessed. I think a fix is out of scope for this PR, especially seeing as the bug likely exists in ow_simulator.

**Comment by AstroStucky** [comment]
> > > I think I found the bug.
> > > OwInterface.cpp:1210
> 
> The same silly bug occurs for the Guarded version of this action on line 1243. I suggest fixing these on this branch -- either of us can commit it.

Good catch! I missed that one. 

A separate hotfix PR on ow_autonomy might be a little easier. We do not want to wrap up the changes on this branch into a hotfix, and I like this approach better than cherry-picking a commit directly into master. 

I'm testing a hotfix branch right now. I'll have a PR posted in a bit. After its merged I can just merge noetic-devel (along with the master changes) back into this one

**Comment by AstroStucky** [comment]
Fixing ReferenceMission1 may be out of the scoop of this PR. 

The reason for these failures is the length of the trench is too short for a **TaskScoopCircular** that occurs at the same position. The errors are not caused by a change in code, but rather are now being reported due to a change in code. That change was https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1090, which was a hotfix, so these errors are likely sometimes occurring in the master branch too. 

I spent some time tinkering with the trench length parameter in ReferenceMission1 to try and increase it enough to incorporate the **TaskScoopCircular**, but when I increase it too much there is the risk that planning for TaskGrind will fail because the far end of the trench winds up outside the arm's reach. Part of the reason this will only happens sometimes is the point selected by IdentifySampleTarget can vary slightly from run to run.

This is off-topic for this PR, but when Plexil has the ability to know whether an action failed, particularly whether a planning error occurred or not, re-attempting planning for a sample a point closer to the lander would be a nice contingency demo. 

I could replicate the **TaskDiscardSample** failure only once. It is a planning failure. I think this requires its own investigation and cannot be fixed in ow_autonomy. Made a ticket for this https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1160

**Comment by Lumsterling** [comment]
> The plan `TestArmMoveCartesianGuarded` has a number of failures and (@Lumsterling) also exposes the excessive power warning:
> ...
> I witnessed exactly when the first excessive power warning occurred: it was when the arm was stuck in the ground, and this was due to a wanky orientation and position of the arm. Apparently the excessive power draw continued even after the arm was unstuck, and stowing.

Sorry, I saw this a few days ago, and then forgot to reply. Sounds like some other issues were discovered, but regarding the power draw specifically:

Keep in mind the power draw that is applied is a moving average of the last 25 raw mechanical power calculations. As such, it's expected that the excessive draw continues after the arm is no longer in an odd position; I'd expect it to continue for 25 cycles after the moment of high draw. 60-80W as an average is still really high though, pretty sure that's not ideal.

**Comment by AstroStucky** [comment]
Went through all Test*.plp plans to make sure they work fine on this branch. All succeeded except the following

_TestActions.plp_ - **DockIngestSample** caused an indefinite block at the end of the plan, similar to what @kmdalal saw. This is already tracked by a Jira, so I won't try to fix it in this branch. Learned some more about what might be going wrong.
_TestArmMoveCartesianGuarded.plp_ - Was failing it turns out it needed the same changes as have been applied to TestActions.plp, this has been done, so I have added it to the tests for this PR. 
_TestPanTiltCartesian.plp_ - The coordinates were incorrect. This too has been fixed in this branch, so this test has been added to the tests for the PR. 

As of now, all Test*.plp plans all work fine, with the exception of an occasional failure of _TestActions.plp_ due to the **DockIngestSample** bug.

**Comment by AstroStucky** [comment]
Ran `ReferenceMission1.plp` once again. All worked fine except for `j_dist_pitch goal error` occurring twice. This will be fixed separately.

**Comment by kmdalal** [comment]
`TestPanTiltMoveCartesian` repeatedly fails:
```
[INFO]: PanTiltMoveCartesian action started
[ERROR]: PanTiltMoveCartesian: Aborted - Requested tilt 1.5644910165286037 is not within allowed limits.
```

**Comment by AstroStucky** [comment]
> `TestPanTiltMoveCartesian` repeatedly fails:
> 

See the comments in that plp file. One of the command calls is intended to fail.

**Comment by kmdalal** [comment]
> > `TestPanTiltMoveCartesian` repeatedly fails:
> 
> See the comments in that plp file. One of the command calls is intended to fail.

I'm not sure why a failure case should be included in the tests, unless that failure triggered some _desired_ behavior such as fault handling.  I suggest removing this call.  If you had a strong reason for including it, a ROS message about the expected failure would be helpful.

File: ow_plexil/src/plans/TestPanTiltMoveCartesian.plp [hunk 1]
```diff
@@ -16,5 +16,5 @@ TestPanTiltMoveCartesian:
   LibraryCall PanTiltMoveCartesian (Frame=BASE_LINK, X=1.46, Y=-0.4, Z=-0.05);
 
   // Point at Jupiter
```
Comment (kmdalal) [inline ow_plexil/src/plans/TestPanTiltMoveCartesian.plp:18]: These comments would better be replaced by ROS messages saying what the antenna is pointing to.

**Comment by AstroStucky** [comment]
This PR was to fix a bug introduced by a change in several action interfaces, not to alter test script behavior.

While I am not sure, I do worry larger changes to test scripts here will become merge conflicts once Release 11.1 is merged in, requiring us to just redo all this work, which is why I wanted to keep this PR small. 

Changes here should only be towards fixing tests currently broken in noetic-devel.

File: ow_plexil/src/plans/TestPanTiltMoveCartesian.plp (hunk 1 above)
Comment (AstroStucky) [reply ow_plexil/src/plans/TestPanTiltMoveCartesian.plp:18]: I don't think we should change test plan behavior in this PR. Please see my most recent thread comment.

**Review by kmdalal — APPROVED** [review]
Got it on not changing test behavior, and keeping the PR small.  I'm still surprised that the previous code had a deliberately failing antenna test, but this can all be sorted out in Release 11.1.

**Pull request merged**